src/dragonball/src/device_manager/blk_dev_mgr.rs

```diff
@@ -0,0 +1,769 @@
+            epoll_mgr,
+            limiters,
+        )?))
+    }
+
+    /// Generated guest kernel commandline related to root block device.
+    pub fn generate_kernel_boot_args(
+        &self,
+        kernel_config: &mut KernelConfigInfo,
+    ) -> std::result::Result<(), DeviceMgrError> {
+        // Respect user configuration if kernel_cmdline contains "root=",
+        // special attention for the case when kernel command line starting with "root=xxx"
+        let old_kernel_cmdline = format!(" {}", kernel_config.kernel_cmdline().as_str());
+        if !old_kernel_cmdline.contains(" root=") && self.has_root_block {
+            let cmdline = kernel_config.kernel_cmdline_mut();
+            if let Some(ref uuid) = self.partuuid {
+                cmdline
+                    .insert("root", &format!("PARTUUID={}", uuid))
+                    .map_err(DeviceMgrError::Cmdline)?;
+            } else {
+                cmdline
+                    .insert("root", "/dev/vda")
+                    .map_err(DeviceMgrError::Cmdline)?;
+            }
+            if self.read_only_root {
+                cmdline.insert_str("ro").map_err(DeviceMgrError::Cmdline)?;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// insert a block device's config. return index on success.
+    fn create(
+        &mut self,
+        block_device_config: BlockDeviceConfigInfo,
+    ) -> std::result::Result<usize, BlockDeviceError> {
+        self.check_data_file_present(&block_device_config)?;
+        if self
+            .get_index_of_drive_path(&block_device_config.path_on_host)
+            .is_some()
+        {
+            return Err(BlockDeviceError::BlockDevicePathAlreadyExists(
+                block_device_config.path_on_host,
+            ));
+        }
+
+        let mut index = 0;
+        // check whether the Device Config belongs to a root device
+        // we need to satisfy the condition by which a VMM can only have on root device
+        if block_device_config.is_root_device {
+            if self.has_root_block {
+                return Err(BlockDeviceError::RootBlockDeviceAlreadyAdded);
+            } else {
+                self.has_root_block = true;
+                self.read_only_root = block_device_config.is_read_only;
+                self.has_partuuid_root = block_device_config.partuuid.is_some();
+                self.partuuid = block_device_config.partuuid.clone();
+                // Root Device should be the first in the list whether or not PARTUUID is specified
+                // in order to avoid bugs in case of switching from partuuid boot scenarios to
+                // /dev/vda boot type.
+                self.info_list
+                    .push_front(BlockDeviceInfo::new(block_device_config));
+            }
+        } else {
+            self.info_list
+                .push_back(BlockDeviceInfo::new(block_device_config));
+            index = self.info_list.len() - 1;
+        }
+
+        Ok(index)
```

Comment:
This code looks a bit odd to me: why not just do this?:

```diff
diff --git a/src/dragonball/src/device_manager/blk_dev_mgr.rs b/src/dragonball/src/device_manager/blk_dev_mgr.rs
index 3afab9c63..b47e3c2df 100644
--- a/src/dragonball/src/device_manager/blk_dev_mgr.rs
+++ b/src/dragonball/src/device_manager/blk_dev_mgr.rs
@@ -612,7 +612,6 @@ impl BlockDeviceMgr {
             ));
         }

-        let mut index = 0;
         // check whether the Device Config belongs to a root device
         // we need to satisfy the condition by which a VMM can only have on root device
         if block_device_config.is_root_device {
@@ -628,14 +627,13 @@ impl BlockDeviceMgr {
                 // /dev/vda boot type.
                 self.info_list
                     .push_front(BlockDeviceInfo::new(block_device_config));
+                Ok(0)
             }
         } else {
             self.info_list
                 .push_back(BlockDeviceInfo::new(block_device_config));
-            index = self.info_list.len() - 1;
+            Ok(self.info_list.len() - 1)
         }
-
-        Ok(index)
     }

     /// Updates a Block Device Config. The update fails if it would result in two
```

Reply:
That looks clearer!!
Thanks, changed.